src/bindings/jna/libelektra/src/main/java/org/libelektra/KDB.java

```diff
@@ -67,52 +41,37 @@
 	 *                                  specialization of {@link KDBException}
 	 * @throws KeySetReleasedException  if {@code contract} has already been
 	 *                                  released
-	 * @throws IllegalArgumentException if {@code contract} or {@code errorKey} is
-	 *                                  {@code null}
+	 * @throws IllegalArgumentException if {@code contract} is {@code null}
+	 * @see #open()
 	 */
 	@Nonnull public static KDB open (KeySet contract) throws KDBException
 	{
 		argNotNull (contract, "KeySet 'contract'");
-		var errorKey = Key.createNameless ();
-		var session = open (contract, errorKey);
-
-		// errorKey is being released if no KDBException occurred
-		errorKey.release ();
-
-		return session;
+		return openInternal (contract);
 	}
 
 	/**
-	 * Opens KDB session using the specified {@code errorKey} to store possible
-	 * warnings and error information
+	 * Opens KDB session
 	 *
-	 * @param contract Contract that will be ensured by
+	 * @param contract Optional contract that will be ensured by
 	 *                 {@link Elektra#kdbOpen(Pointer, Pointer)}
-	 * @param errorKey Used to store warnings and error information
 	 * @return New KDB session
-	 * @throws KDBException             if opening the session fails - see
-	 *                                  specialization of {@link KDBException}
-	 * @throws KeySetReleasedException  if {@code contract} has already been
-	 *                                  released
-	 * @throws KeyReleasedException     if {@code errorKey} has already been
-	 *                                  released
-	 * @throws IllegalArgumentException if {@code contract} or {@code errorKey} is
-	 *                                  {@code null}
-	 * @see Key#createNameless()
+	 * @throws KDBException            if opening the session fails - see
+	 *                                 specialization of {@link KDBException}
+	 * @throws KeySetReleasedException if {@code contract} has already been released
 	 */
-	@Nonnull public static KDB open (KeySet contract, Key errorKey) throws KDBException
-	{
-		argNotNull (contract, "KeySet 'contract'");
-		argNotNull (errorKey, "Key 'errorKey'");
-		return checkKDBPointer (Elektra.INSTANCE.kdbOpen (contract.getPointer (), errorKey.getPointer ()), errorKey);
-	}
-
-	@Nonnull private static KDB checkKDBPointer (@Nullable Pointer pointer, Key errorKey) throws KDBException
+	@Nonnull private static KDB openInternal (@Nullable KeySet contract) throws KDBException
```

Comment:
Why don't we just allow `KDB.open(null)`?

Reply:
Not permitting NULL valued arguments in a (public) Java APIs is considered good practice in my professional experience. If you want to provide a `contract`, use the signature providing the opportunity to do so. Otherwise use the signature without the `contract` argument. Since Java does not support optional method arguments, it is common to provide different signatures for different combinations of valid arguments. For more than two optional parameters, Builders are commonly expected to be provided.
Together with static code analysis in conjunction with the relevant annotations, this should lead to a better developer experience, less errors and better code readability.

Reply:
I get that the `KDB.open()` overload without arguments is useful, but why have the additional copy of `KDB.open(KeySet)`?

This just makes it more difficult, if you want to wrap `KDB.open` in another function and keep the contract optional. You either need to also create two overloads, or you have to add an useless `if` to call the correct overload when both overloads will call the same implementation anyway.

Take this hypothetical wrapper for example:
```java
// we need to do this
KDB openWrapper(KeySet ​contract) {
    KDB kdb;
    if (contract == null) {
        kdb = KDB.open();
    } else {
        kdb = KDB.open(contract);
    }
    // do something
    return kdb;
​}

// when this would also be an option
KDB openWrapper(KeySet contract) {
   ​KDB kdb = KDB.open(contract);
   ​// do something
   ​return kdb;
}
```

Note: Using `Optional<KeySet>` can make this a bit more readable, but we still have the useless branching and then executing the same thing anyway.

Reply:
Correct. This is always the problem to consider, in cases when having too few optional parameters to sensibly opt for a builder approach. Regarding the decision for whether to introduce nullable parameters and therefore weaken means of static  analysis of potential null pointer problems, i would opt for going with the separate signatures.
I'm not sure how important (= frequent) the open wrapper use case example is, but one can always write it in a readable way as a one liner using a trinary:
```
@Nonull KDB openWrapper(@Nullable KeySet contract) {
   try (​KDB kdb = contract == null ? KDB.open() : KDB.open(contract)) {
     ​// do something
     ​return kdb;
  }
}
```

However we decide we should do so by establishing a rule, when to allow null valued parameters in the public Java binding API.

Reply:
As a general rule i would propose not allowing null valued arguments in the public Java binding except for significant use cases. This should then be pointed out (emphasized) in the corresponding method java doc.